libcni/api.go

```diff
@@ -336,6 +339,85 @@ func (c *CNIConfig) DelNetwork(net *NetworkConfig, rt *RuntimeConf) error {
 	return nil
 }
 
+// ValidateNetworkList checks that a configuration is reasonably valid.
+// - all the specified plugins exist on disk
+// - every plugin supports the desired version.
+//
+// It returns a list of missing but desired capabilities and error if the configuration
+// is not executable. Runtimes should decide if they care about the missing capabilities
+// separately
+func (c *CNIConfig) ValidateNetworkList(list *NetworkConfigList, desiredCapabilities []string) ([]string, error) {
+	version := list.CNIVersion
+
+	missingCaps := map[string]interface{}{}
+	for _, cap := range desiredCapabilities {
+		missingCaps[cap] = nil
+	}
+
+	errs := []error{}
+	for _, net := range list.Plugins {
+		if err := c.validatePlugin(net.Network.Type, version); err != nil {
+			errs = append(errs, err)
+		}
+		for cap, enabled := range net.Network.Capabilities {
+			if !enabled {
+				continue
+			}
+
+			if missingCaps != nil {
+				if _, ok := missingCaps[cap]; ok {
+					delete(missingCaps, cap)
+				}
+			}
+		}
+	}
+
+	// scan for missing capabilities
+	caps := []string{}
+	if len(missingCaps) != 0 {
```

Comment:
don't need this len() != 0 either, the range will just not execute if there aren't any missingCaps.